etc/ci/upstream-wpt-changes/wptupstreamer/__init__.py

```diff
@@ -0,0 +1,260 @@
+# Copyright 2023 The Servo Project Developers. See the COPYRIGHT
+# file at the top-level directory of this distribution.
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+# <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+
+# pylint: disable=broad-except
+# pylint: disable=dangerous-default-value
+# pylint: disable=fixme
+# pylint: disable=missing-docstring
+
+# This allows using types that are defined later in the file.
+from __future__ import annotations
+
+import dataclasses
+import json
+import logging
+import re
+import subprocess
+
+from typing import Callable, Optional
+
+from .common import \
+    CLOSING_EXISTING_UPSTREAM_PR, \
+    NO_SYNC_SIGNAL, \
+    NO_UPSTREAMBLE_CHANGES_COMMENT, \
+    OPENED_NEW_UPSTREAM_PR, \
+    UPDATED_EXISTING_UPSTREAM_PR, \
+    UPDATED_TITLE_IN_EXISTING_UPSTREAM_PR, \
+    UPSTREAMABLE_PATH, \
+    wpt_branch_name_from_servo_pr_number
+
+from .github import GithubRepository, PullRequest
+from .step import \
+    AsyncValue, \
+    ChangePRStep, \
+    CommentStep, \
+    CreateOrUpdateBranchForPRStep, \
+    MergePRStep, \
+    OpenPRStep, \
+    RemoveBranchForPRStep, \
+    Step
+
+
+class LocalGitRepo:
+    def __init__(self, path: str, sync: WPTSync):
+        self.path = path
+        self.sync = sync
+
+    def run(self, *args, env: dict = {}):
+        command_line = ["git"] + list(args)
+        logging.info("  → Execution (cwd='%s'): %s",
+                     self.path, " ".join(command_line))
+
+        env.setdefault("GIT_AUTHOR_EMAIL", self.sync.github_email)
+        env.setdefault("GIT_COMMITTER_EMAIL", self.sync.github_email)
+        env.setdefault("GIT_AUTHOR_NAME", self.sync.github_name)
+        env.setdefault("GIT_COMMITTER_NAME", self.sync.github_name)
+
+        return subprocess.check_output(
+            command_line, cwd=self.path, env=env, stderr=subprocess.STDOUT
+        ).decode("utf-8")
+
+
+@dataclasses.dataclass()
+class SyncRun:
+    sync: WPTSync
+    servo_pr: PullRequest
+    upstream_pr: Optional[PullRequest]
+    step_callback: Optional[Callable[[Step], None]]
+    steps: list[Step] = dataclasses.field(default_factory=list)
+
+    def make_comment(self, template: str) -> str:
+        return template.format(
+            upstream_pr=self.upstream_pr,
+            servo_pr=self.servo_pr,
+        )
+
+    def add_step(self, step) -> Optional[AsyncValue]:
+        self.steps.append(step)
+        return step.provides()
+
+    def run(self):
+        while self.steps:
+            step = self.steps.pop(0)
```

Comment:
Thanks. I've added a comment here.

In addition, I've also made it so that the `AsyncValue` produced by `OpenPRStep` is used to fill out the `SyncRun.upstream_pr` member instead of having the step set this value directly. This makes `OpenPR` step a bit more general.